### PR TITLE
[ OPSRC-92 ] Update Sylius version to 1.9 and resolve issues

### LINKED ISSUE #16: 404 Resource not found
Hi,

If you got this error when entering a new email address  :

> 404
> ResourceNotFound
> The requested resource could not be found.
> 
> This error tells you a specific resource doesn’t exist. It’s possible that the resource has been moved or deleted, or that there’s a typo in your request.

Don't forget to add quotes to your API KEY and LIST ID.
It'll escape you to loose one full morning.

PS: you should edit the readme for the new version of Sylius (1.3) with Symfony 4 (it's change lot of things about the installation)

### LINKED ISSUE #28: Synchronization with MailChimp's list of users
When a user unsubscribes using the link from MailChimp's template, it is not synced with the sylius project (the user still appears as subscribed in the database).

### LINKED ISSUE #34: Updating email for customer subscribed to newsletter results in UniqueConstraintViolationException
If editing a customer's email address through the admin interface and the customer is subscribed to the newsletter, the `NewsletterSubscriptionHandler` performs changes that results in a unique constraint violation because it is trying to create a new customer record with the updated email address.

```sh
Doctrine\DBAL\Exception\UniqueConstraintViolationException: An exception occurred while executing 'UPDATE sylius_customer SET email = ?, email_canonical = ?, updated_at = ? WHERE id = ?' with params ["test-email@example.com", "test-email@example.com", "2020-09-28 17:55:43", 3781]:

SQLSTATE[23000]: Integrity constraint violation: 1062 Duplicate entry 'test-email@example.com' for key 'UNIQ_7E82D5E6E7927C74'
#25 /vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php(59): Doctrine\DBAL\Driver\AbstractMySQLDriver::convertException
#24 /vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php(172): Doctrine\DBAL\DBALException::wrapException
#23 /vendor/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php(151): Doctrine\DBAL\DBALException::driverExceptionDuringQuery
#22 /vendor/doctrine/dbal/lib/Doctrine/DBAL/Connection.php(1087): Doctrine\DBAL\Connection::executeUpdate
#21 /vendor/doctrine/orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php(500): Doctrine\ORM\Persisters\Entity\BasicEntityPersister::updateTable
#20 /vendor/doctrine/orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php(389): Doctrine\ORM\Persisters\Entity\BasicEntityPersister::update
#19 /vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php(1171): Doctrine\ORM\UnitOfWork::executeUpdates
#18 /vendor/doctrine/orm/lib/Doctrine/ORM/UnitOfWork.php(406): Doctrine\ORM\UnitOfWork::commit
#17 /vendor/doctrine/orm/lib/Doctrine/ORM/EntityManager.php(370): Doctrine\ORM\EntityManager::flush
#16 /var/cache/prod/ContainerBrx3oPf/EntityManager_9a5be93.php(86): EntityManager_9a5be93::flush
#15 /vendor/sylius/resource-bundle/src/Bundle/Doctrine/ORM/EntityRepository.php(32): Sylius\Bundle\ResourceBundle\Doctrine\ORM\EntityRepository::add
#14 /vendor/bitbag/mailchimp-plugin/src/Handler/NewsletterSubscriptionHandler.php(108): BitBag\SyliusMailChimpPlugin\Handler\NewsletterSubscriptionHandler::createNewCustomer
#13 /vendor/bitbag/mailchimp-plugin/src/Handler/NewsletterSubscriptionHandler.php(76): BitBag\SyliusMailChimpPlugin\Handler\NewsletterSubscriptionHandler::subscribe
#12 /vendor/bitbag/mailchimp-plugin/src/EventListener/CustomerNewsletterListener.php(55): BitBag\SyliusMailChimpPlugin\EventListener\CustomerNewsletterListener::subscribe
#11 /vendor/bitbag/mailchimp-plugin/src/EventListener/CustomerNewsletterListener.php(42): BitBag\SyliusMailChimpPlugin\EventListener\CustomerNewsletterListener::customerCreateEvent
#10 /vendor/bitbag/mailchimp-plugin/src/EventListener/CustomerNewsletterListener.php(50): BitBag\SyliusMailChimpPlugin\EventListener\CustomerNewsletterListener::customerUpdateEvent
#9 /vendor/symfony/event-dispatcher/EventDispatcher.php(304): Symfony\Component\EventDispatcher\EventDispatcher::Symfony\Component\EventDispatcher\{closure}
#8 /vendor/symfony/event-dispatcher/EventDispatcher.php(264): Symfony\Component\EventDispatcher\EventDispatcher::doDispatch
#7 /vendor/symfony/event-dispatcher/EventDispatcher.php(239): Symfony\Component\EventDispatcher\EventDispatcher::callListeners
#6 /vendor/symfony/event-dispatcher/EventDispatcher.php(73): Symfony\Component\EventDispatcher\EventDispatcher::dispatch
#5 /vendor/sylius/resource-bundle/src/Bundle/Controller/EventDispatcher.php(76): Sylius\Bundle\ResourceBundle\Controller\EventDispatcher::dispatchPreEvent
#4 /vendor/sylius/resource-bundle/src/Bundle/Controller/ResourceController.php(268): Sylius\Bundle\ResourceBundle\Controller\ResourceController::updateAction
#3 /vendor/symfony/http-kernel/HttpKernel.php(158): Symfony\Component\HttpKernel\HttpKernel::handleRaw
#2 /vendor/symfony/http-kernel/HttpKernel.php(80): Symfony\Component\HttpKernel\HttpKernel::handle
#1 /vendor/symfony/http-kernel/Kernel.php(201): Symfony\Component\HttpKernel\Kernel::handle
#0 /public/index.php(25): null
```

### PR DESCRIPTION
**Pull request by tbuczen**
Fixed following issues:
https://github.com/BitBagCommerce/SyliusMailChimpPlugin/issues/34
https://github.com/BitBagCommerce/SyliusMailChimpPlugin/issues/16

Added suggested enahncement:
https://github.com/BitBagCommerce/SyliusMailChimpPlugin/issues/28

### CODE REVIEW

File: tests/Application/.env.test [hunk 1]
```diff
@@ -4,6 +4,6 @@ KERNEL_CLASS='Tests\BitBag\SyliusExamplePlugin\Application\Kernel'
 
 DATABASE_URL=sqlite:///%kernel.project_dir%/var/data.db
 
-MAIL_CHIMP_API_KEY=6a74ec890676264ec8116a7945929d12-us18
-
-MAIL_CHIMP_LIST_ID=8af647b5c0
+MAIL_CHIMP_API_KEY='5cf6212ae1aa1b465f5c53124b9afb54-us1'
+MAIL_CHIMP_LIST_ID='237417b17d'
+MAIL_CHIMP_WEBHOOK_SECRET='237417b17d'
```
Comment (xNordo) [inline tests/Application/.env.test]: Should this be commited?

File: ecs.php [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+declare(strict_types=1);
+
+use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
+
+return static function (ContainerConfigurator $containerConfigurator): void {
+    $containerConfigurator->import('vendor/sylius-labs/coding-standard/ecs.php');
+};
```
Comment (damonsson) [inline ecs.php]: Missing empty line at the end of the file

https://stackoverflow.com/questions/36019534/add-an-empty-line-at-end-of-file-according-to-psr-2-on-phpstorm

File: tests/Application/.env.test (hunk 1 above)
Comment (tbuczen) [reply tests/Application/.env.test]: Good point, Mailchimp would disable any keys that are publicly visible, I'm going to change it to dummy text

File: src/BitBagSyliusMailChimpPlugin.php [hunk 3]
```diff
@@ -1,5 +1,13 @@
 <?php
 
+/*
+ * This file has been created by developers from BitBag.
+ * Feel free to contact us once you face any issues or want to start
+ * another great project.
+ * You can find more information about us on https://bitbag.shop and write us
+ * an email on tomasz.grochowski@bitbag.pl.
+ */
+
```
Comment (damonsson) [inline src/BitBagSyliusMailChimpPlugin.php]: We have new text for this IIRC, please ask Krystian about that. And trust me, you don't want to leave your email here 🙉

File: src/Controller/MailchimpController.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
+<?php
+
+/*
+ * This file has been created by developers from BitBag.
+ * Feel free to contact us once you face any issues or want to start
+ * another great project.
+ * You can find more information about us on https://bitbag.shop and write us
+ * an email on tomasz.grochowski@bitbag.pl.
+ */
+
```
Comment (damonsson) [inline src/Controller/MailchimpController.php]: We have new text for this IIRC, please ask Krystian about that. And trust me, you don't want to leave your email here 🙉

File: src/Controller/NewsletterController.php [hunk 5]
```diff
@@ -1,37 +1,73 @@
 <?php
 
+/*
+ * This file has been created by developers from BitBag.
+ * Feel free to contact us once you face any issues or want to start
+ * another great project.
+ * You can find more information about us on https://bitbag.shop and write us
+ * an email on tomasz.grochowski@bitbag.pl.
+ */
+
```
Comment (damonsson) [inline src/Controller/NewsletterController.php]: We have new text for this IIRC, please ask Krystian about that. And trust me, you don't want to leave your email here 🙉

File: src/DependencyInjection/BitBagSyliusMailChimpExtension.php [hunk 6]
```diff
@@ -1,5 +1,13 @@
 <?php
 
+/*
+ * This file has been created by developers from BitBag.
+ * Feel free to contact us once you face any issues or want to start
+ * another great project.
+ * You can find more information about us on https://bitbag.shop and write us
+ * an email on tomasz.grochowski@bitbag.pl.
+ */
+
```
Comment (damonsson) [inline src/DependencyInjection/BitBagSyliusMailChimpExtension.php]: We have new text for this IIRC, please ask Krystian about that. And trust me, you don't want to leave your email here 🙉

File: src/EventListener/CustomerNewsletterListener.php [hunk 7]
```diff
@@ -1,35 +1,39 @@
 <?php
 
+/*
+ * This file has been created by developers from BitBag.
+ * Feel free to contact us once you face any issues or want to start
+ * another great project.
+ * You can find more information about us on https://bitbag.shop and write us
+ * an email on tomasz.grochowski@bitbag.pl.
+ */
+
```
Comment (damonsson) [inline src/EventListener/CustomerNewsletterListener.php]: We have new text for this IIRC, please ask Krystian about that. And trust me, you don't want to leave your email here 🙉

File: src/EventListener/CustomerNewsletterListener.php [hunk 8]
```diff
@@ -42,21 +46,59 @@ public function customerCreateEvent(GenericEvent $event)
...
+        $customer = $event->getSubject();
+
+        if (!$customer instanceof CustomerInterface) {
+            throw new UnexpectedTypeException(
+                $customer,
+                CustomerInterface::class
+            );
+        }
+
+        $uow = $this->entityManager->getUnitOfWork();
```
Comment (damonsson) [inline src/EventListener/CustomerNewsletterListener.php]: What do you think about full $unitOfWork, instead shortcut $uow?

File: src/Handler/NewsletterSubscriptionHandler.php [hunk 9]
```diff
@@ -1,5 +1,13 @@
 <?php
 
+/*
+ * This file has been created by developers from BitBag.
+ * Feel free to contact us once you face any issues or want to start
+ * another great project.
+ * You can find more information about us on https://bitbag.shop and write us
+ * an email on tomasz.grochowski@bitbag.pl.
+ */
+
```
Comment (damonsson) [inline src/Handler/NewsletterSubscriptionHandler.php]: We have new text for this IIRC, please ask Krystian about that. And trust me, you don't want to leave your email here 🙉

File: src/Handler/NewsletterSubscriptionHandler.php [hunk 10]
```diff
@@ -63,85 +53,134 @@ public function __construct(
...
 
-        if ($response['status'] === Response::HTTP_NOT_FOUND) {
-            $this->exportNewEmail($email);
+        $customer->setSubscribedToNewsletter(true);
+        $this->customerManager->flush();
+    }
+
+    public function getValidMailchimpListIds(): array
+    {
+        $responseArray = $this->mailChimp->get('/lists');
```
Comment (damonsson) [inline src/Handler/NewsletterSubscriptionHandler.php]: Maybe move this hardcoded strings like "lists" to consts? Looks like we are using it more than once.

File: src/Model/WebhookData.php [hunk 11]
```diff
@@ -0,0 +1,76 @@
+<?php
+
+/*
+ * This file has been created by developers from BitBag.
+ * Feel free to contact us once you face any issues or want to start
+ * another great project.
+ * You can find more information about us on https://bitbag.shop and write us
+ * an email on tomasz.grochowski@bitbag.pl.
+ */
```
Comment (damonsson) [inline src/Model/WebhookData.php]: We have new text for this IIRC, please ask Krystian about that. And trust me, you don't want to leave your email here 🙉

File: src/Resources/translations/messages.en.yml [hunk 12]
```diff
@@ -5,3 +5,4 @@ bitbag_sylius_mailchimp_plugin:
         your_email_address: Your email address
         subscribed_successfully: You are now subscribed to the newsletter.
         invalid_csrf_token: Submited CSRF token is invalid.
+        unsubscribed_successfully: You are now unsubscribed from the newsletter.
```
Comment (damonsson) [inline src/Resources/translations/messages.en.yml]: empty line missing

File: src/Validator/Constraints/UniqueNewsletterEmail.php [hunk 13]
```diff
@@ -1,5 +1,13 @@
 <?php
 
+/*
+ * This file has been created by developers from BitBag.
+ * Feel free to contact us once you face any issues or want to start
+ * another great project.
+ * You can find more information about us on https://bitbag.shop and write us
+ * an email on tomasz.grochowski@bitbag.pl.
+ */
+
```
Comment (damonsson) [inline src/Validator/Constraints/UniqueNewsletterEmail.php]: We have new text for this IIRC, please ask Krystian about that. And trust me, you don't want to leave your email here 🙉

File: src/Validator/Constraints/UniqueNewsletterEmailValidator.php [hunk 14]
```diff
@@ -1,5 +1,13 @@
 <?php
 
+/*
+ * This file has been created by developers from BitBag.
+ * Feel free to contact us once you face any issues or want to start
+ * another great project.
+ * You can find more information about us on https://bitbag.shop and write us
+ * an email on tomasz.grochowski@bitbag.pl.
+ */
+
```
Comment (damonsson) [inline src/Validator/Constraints/UniqueNewsletterEmailValidator.php]: We have new text for this IIRC, please ask Krystian about that. And trust me, you don't want to leave your email here 🙉

File: src/Validator/NewsletterValidator.php [hunk 15]
```diff
@@ -1,44 +1,34 @@
 <?php
 
+/*
+ * This file has been created by developers from BitBag.
+ * Feel free to contact us once you face any issues or want to start
+ * another great project.
+ * You can find more information about us on https://bitbag.shop and write us
+ * an email on tomasz.grochowski@bitbag.pl.
+ */
+
```
Comment (damonsson) [inline src/Validator/NewsletterValidator.php]: We have new text for this IIRC, please ask Krystian about that. And trust me, you don't want to leave your email here 🙉

File: src/Validator/WebhookValidator.php [hunk 16]
```diff
@@ -0,0 +1,92 @@
+<?php
+
+/*
+ * This file has been created by developers from BitBag.
+ * Feel free to contact us once you face any issues or want to start
+ * another great project.
+ * You can find more information about us on https://bitbag.shop and write us
+ * an email on tomasz.grochowski@bitbag.pl.
+ */
```
Comment (damonsson) [inline src/Validator/WebhookValidator.php]: We have new text for this IIRC, please ask Krystian about that. And trust me, you don't want to leave your email here 🙉

File: tests/Application/config/packages/security.yaml [hunk 17]
```diff
@@ -1,16 +1,27 @@
...
             id: sylius.shop_user_provider.email_or_name_based
+        sylius_api_shop_user_provider:
+            id: sylius.shop_user_provider.email_or_name_based
+        sylius_api_chain_provider:
+            chain:
+                providers: [sylius_api_shop_user_provider, sylius_api_admin_user_provider]
+
     encoders:
-        Sylius\Component\User\Model\UserInterface: sha512
+        Sylius\Component\User\Model\UserInterface: argon2i
```
Comment (damonsson) [inline tests/Application/config/packages/security.yaml]: argon2i in tests? What about speed up?

File: tests/Behat/Resources/suites.yml [hunk 18]
```diff
@@ -75,3 +75,16 @@ default:
...
+                - sylius.behat.context.setup.channel
+                - sylius.behat.context.setup.customer
+                - sylius.behat.context.setup.admin_security
+                - sylius.behat.context.ui.admin.managing_customers
+
+                - bitbag_sylius_mailchimp_plugin.context.ui.shop.mailchimp
+                - bitbag_sylius_mailchimp_plugin.context.ui.shop.newsletter
+
+            filters:
+                tags: "@admin && @ui"
```
Comment (damonsson) [inline tests/Behat/Resources/suites.yml]: missing empty line at the end of file

File: .github/workflows/build.yml [hunk 19]
```diff
@@ -0,0 +1,208 @@
...
+                run: bin/console doctrine:schema:validate
+
+            -
+                name: Run PHPSpec
+                run: vendor/bin/phpspec run --ansi -f progress --no-interaction
+
+            -
+                name: Run PHPUnit
+                run: vendor/bin/phpunit --colors=always
+#Temporary disabled
```
Comment (leszczuu) [inline .github/workflows/build.yml]: lets remove it, We can add it later when its needed

File: .scrutinizer.yml [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+    analysis:
+      tests:
+        override:
+          - php-scrutinizer-run
+  environment:
+    variables:
+      COMPOSER_MEMORY_LIMIT: -1
+
+filter:
+  excluded_paths: [tests/*, spec/*]
```
Comment (leszczuu) [inline .scrutinizer.yml]: fix 'no new line at the end' errors

**Review by leszczuu — APPROVED** [review]

**Review by damonsson — APPROVED** [review]

**Pull request merged**